packages/storybook/stories/QuickStarts/snippets/QuickstartCompositeAdapter.snippet.tsx

```diff
@@ -1,70 +1,94 @@
-  const [callAdapter, setCallAdapter] = useState<CallAdapter>();
+// Fill in the following configuration inline for this sample.
+// In a real application, these must be supplied securely to the client application.
+const DISPLAY_NAME = '<Display Name>';
+const ENDPOINT_URL = '<Azure Communication Services Resource Endpoint>';
+const TOKEN = '<Azure Communication Services Access Token>';
+const USER_ID = '<Azure Communication Services Identifier>';
+// CallComposite specific configuration:
+// For Group Id, developers can pass any GUID they can generate
+const GROUP_ID = '<Developer generated GUID>';
+// ChatComposite specific configuration:
+const THREAD_ID = '<Get thread id from chat service>';
 
-  //Chat Variables
-  const threadId = '<Get thread id from chat service>';
-  const [chatAdapter, setChatAdapter] = useState<ChatAdapter>();
-
-  // We can't even initialize the Chat and Call adapters without a well-formed token.
+function App(): JSX.Element {
+  // CallComposite and ChatComposite need a well-formed token for initialization.
   const credential = useMemo(() => {
     try {
-      return new AzureCommunicationTokenCredential(token);
+      return new AzureCommunicationTokenCredential(TOKEN);
     } catch {
-      console.error('Failed to construct token credential');
       return undefined;
     }
-  }, [token]);
+  }, []);
+  const callAdapter = useCallAdapter(DISPLAY_NAME, GROUP_ID, USER_ID, credential);
+  const chatAdapter = useChatAdapter(DISPLAY_NAME, ENDPOINT_URL, THREAD_ID, USER_ID, credential);
+
+  if (!credential) {
+    return <h3>Failed to construct credential. Provided token is malformed.</h3>;
+  }
+  if (!callAdapter || !chatAdapter) {
+    return <h3>Initializing...</h3>;
+  }
+  return <h1>Hooray! You set up adapters 🎉🎉🎉</h1>;
+}
 
+function useCallAdapter(
+  displayName: string,
+  groupId: string,
+  userId: string,
+  credential?: AzureCommunicationTokenCredential
+): CallAdapter | undefined {
+  const [callAdapter, setCallAdapter] = useState<CallAdapter>();
   useEffect(() => {
-    if (credential !== undefined) {
-      const createAdapter = async (credential: AzureCommunicationTokenCredential): Promise<void> => {
-        setChatAdapter(
-          await createAzureCommunicationChatAdapter({
-            endpoint: endpointUrl,
-            userId: fromFlatCommunicationIdentifier(userId) as CommunicationUserIdentifier,
-            displayName,
+    (async (): Promise<void> => {
+      credential &&
+        setCallAdapter(
+          await createAzureCommunicationCallAdapter({
             credential,
-            threadId
+            displayName,
+            locator: { groupId },
+            userId: fromFlatCommunicationIdentifier(userId) as CommunicationUserIdentifier
           })
         );
-        setCallAdapter(
-          await createAzureCommunicationCallAdapter({
+    })();
+  }, [credential, displayName, groupId, userId]);
+  return callAdapter;
+}
+
+function useChatAdapter(
```

Comment:
There are a few problems with providing `useChatAdapter` as part of our API:

* `userId` argument is an object. We'd either have to accept flat string `userId` for this method (breaking from the API of `createAzureCommunicationChatAdapter`, or Contoso would have to memoize the passed in object to avoid unnecessary recreation of the adapter.
* `credential` is similarly an object. There is an extra complication with `credential` that Contoso needs to decide what to render if they fail to generate the `credential`. `useChatAdapter` must be called unconditionally (rule of hooks) so it will make for non-trivial logic in Contoso's component to ensure that `credential` is valid before calling the `useChatAdapter` hook.

I'd argue that while it might make sense to extract a local hook for creating the adapter, like here, it's not a good idea to introduce that in our quick-starts. The local custom hook I added in this PR requires the reader to be comfortable with custom hooks, without adding much value (if we provided the hook in our API, that'd be a value-add).

Reply:
@JamesBurnside @PorterNan

Reply:
(Note that we can't accept `token` as a string. In a real application, the provided `credential` should auto-refresh the token as needed).

Reply:
Can we do any internal memorization inside the hook where we explicitly check the contents of the userId and credential objects?

Reply:
I think the value with the shorter code snippets is still worth it. I think that while there is the added complexity of the useAdapter custom hook, there is a removal of complexity about managing the adapter in the snippet (i.e. useStates and useEffects lying around, not having to read about async functions that trigger rerenders).

Code for just chat is super small:
![image](https://user-images.githubusercontent.com/2684369/155438740-f637423f-b583-4cd9-9f2d-4055a9ce73b4.png)

(userId needing memoized in some fashion, either useMemo or we could perhaps internally memoize the userId string)

Would be nice if we could not require the useMemos though